### PR TITLE
Implement vde_vmnet management

### LINKED ISSUE #175: Integrated VDE networking support
I think Lima should manage the VDE networking setup for the user. I would like to make it work like this:

## Lima Network Configuration

Managed networks are declared in `$LIMA_HOME/_config/networks.yaml`. Lima will provide a default configuration:

```yaml
# Paths to vde executables. Because vde_vmnet is invoked via sudo it should be
# installed where only root can modify/replace it. This means also none of the
# parent directories should be writable by the user.
#
# The var_run directory also must not be writable by the user because it will
# include the vde_vmnet pid files. Those will be terminated via sudo, so replacing
# the pid files would allow killing of arbitrary privileged processes.
paths:
  vde_switch: /opt/vde/bin/vde_switch
  vde_vmnet: /opt/vde/bin/vde_vmnet
  var_run: /var/run/lima

group: staff

networks:
  shared:
    mode: shared
    gateway: 192.168.105.1
    dhcp-end: 129.168.105.254
    netmask: 255.255.255.0
  bridged:
    mode: bridged
    interface: en0
    # bridged mode doesn't has a gateway; dhcp is managed by outside network
  host:
    mode: host
    gateway: 192.168.106.1
    dhcp-end: 129.168.106.254
    netmask: 255.255.255.0
```

Users can modify existing networks (e.g. change the dhcp range), or define additional ones.

## Instance Network Configuration

The managed networks can be referenced from an instance's `lima.yaml` using a custom `lima://` VNL scheme:

```
networks:
  vde:
  - vnl: lima://shared
    name: nat
  - vnl: lima://host
```

The default name for networks with a `lima://` VNL will be the network name, not `vdeN`. So the config above will create a `nat` and a `host` interface inside the instance.

Whenever Lima starts an instance, it will check if all referenced networks are already running, and if not, start both `vde_switch` and `vde_vmnet` daemons for that network. Sockets and pid files will be stored under `/var/run/lima`. Log files will be stored under `$LIMA_HOME/_vde`. (Or should we put them into `/var/run/lima` as well? Maybe needs another setting under `paths`.)

When an instance is stopped, Lima will check if there are networks that are no longer being used, and shut them down as well.

## Managing root privileges

`vde_vmnet` has to be run as root, so Lima will start and stop it using `sudo`. This will ask the user for their admin password unless they have enabled password-less `sudo`.

A helper command for more locked-down systems will be implemented to allow Lima to manage networks without asking for the password every time a network is started or stopped:

```console
$ limactl sudoers | sudo tee /etc/sudoers.d/lima
```

This will write the specific commands needed to start and stop the daemons to `/etc/sudoers.d/lima`, e.g.

```
Cmnd_Alias SHARED_SWITCH_START = /opt/vde/bin/vde_switch --sock=/var/run/lima/shared.ctl \
                                 --pidfile=/var/run/lima/shared_switch.pid --group=staff --dirmode=0770
Cmnd_Alias SHARED_SWITCH_STOP  = /usr/bin/pkill -F /var/run/lima/shared_switch.pid

Cmnd_Alias SHARED_VMNET_START  = /opt/vde/bin/vde_vmnet --vmnet-mode=shared /var/run/lima/shared.ctl
Cmnd_Alias SHARED_VMNET_STOP   = /usr/bin/pkill -F /var/run/lima/shared_vmnet.pid

%staff ALL=(root:root)    NOPASSWD:NOSETENV: /bin/mkdir -m 755 -p /var/run/lima
%staff ALL=(daemon:staff) NOPASSWD:NOSETENV: SHARED_SWITCH_START, SHARED_SWITCH_STOP
%staff ALL=(root:root)    NOPASSWD:NOSETENV: SHARED_VMNET_START,  SHARED_VMNET_STOP
```

(Real settings will include `gateway`, `dhcp-end`, `netmask`, etc too. Also `vde_vmnet` doesn't yet support writing a pid file.)

Question: do we really need to run `vde_switch` using the `daemon` user? We run all of `qemu` under the actual user account; is there anything in the switch that makes it more risky than other utilities, so it needs to be segregated?

Every time the `network.yaml` file is edited, the sudoers file must be regenerated.

### PR DESCRIPTION
**Pull request by jandubois**
Lima uses ~/.lima/_config/networks.yaml to define managed networks for instances. By default "shared", "bridged", and "host" are defined. They can be referenced from `lima.yaml` via e.g. `lima://shared`.

The networks will be automatically started by lime as long as any instance referencing them are running, and once the last instance using a network is stopped, the network is brought down as well.

A new command `limactl sudoers` can be used to write a sudoers file to allow lima to start/stop the daemons via `sudo`. Lima does not support prompting for a sudo password; either the sudoers file must
be maintained, or the user must have password-less sudo configured.

Every time the `networks.yaml` configuration is changed, the sudoers file must be regenerated to match.

Resolves #175

### CODE REVIEW

**Comment by jandubois** [comment]
@AkihiroSuda Could you please do a first-pass review?  The PR is fully functional, but I want to add a few things still:

* support for network metrics, so the guest OS will prefer using the `vde` devices over `eth0`.
* some validation for `networks.yaml`
* update documentation

You set up your sudoers file with:

```console
$ limactl sudoers | sudo tee /etc/sudoers.d/lima
```

Sample network declaration in `lima.yaml`:

```yaml
network:
  vde:
  - vnl: lima://shared
```

File: pkg/limayaml/validate.go [hunk 1]
```diff
@@ -177,7 +177,14 @@ func validateNetwork(yNetwork Network) error {
 		// The field is called VDE.VNL in anticipation of QEMU upgrading VDE2 to VDEplug4,
 		// but right now the only valid value on macOS is a path to the vde_switch socket directory,
 		// optionally with vde:// prefix.
-		if !strings.Contains(vde.VNL, "://") || strings.HasPrefix(vde.VNL, "vde://") {
+		// TODO: use networks.LimaSchema after solving circular dependency
+		if strings.HasPrefix(vde.VNL, "lima://") {
```
Comment (AkihiroSuda) [inline pkg/limayaml/validate.go:181]: Sorry, I still think that we shouldn't contaminate the VNL namespace.
`lima://` is reserved for a potential VDE4 plugin that would be called "lima".

https://github.com/lima-vm/lima/issues/175#issuecomment-907589748

File: pkg/networks/commands.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package networks
+
+import (
+	"fmt"
+	"github.com/lima-vm/lima/pkg/store"
+)
+
+const (
+	Switch = "switch"
```
Comment (AkihiroSuda) [inline pkg/networks/commands.go:9]: For performance reason, we may just ditch the switch and use PTP mode everywhere: https://github.com/lima-vm/vde_vmnet/issues/26#issuecomment-917392360

That may also beneficial for simplicity.

File: pkg/networks/commands.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+func (config *NetworksConfig) StartCmd(name, daemon string) string {
+	var cmd string
+	switch daemon {
+	case Switch:
+		cmd = fmt.Sprintf("%s --pidfile=%s --sock=%s --group=%s --dirmode=0770 --nostdin",
+			config.Paths.VDESwitch, config.PIDFile(name, Switch), config.VDESock(name), config.Group)
+	case VMNet:
+		nw := config.Networks[name]
+		cmd = fmt.Sprintf("%s --pidfile=%s --vde-group=%s --vmnet-mode=%s",
```
Comment (AkihiroSuda) [inline pkg/networks/commands.go:68]: Can we error out when the `vde_vmnet` binary (or its DLLs) is modifiable by non-root?

File: cmd/limactl/sudoers.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "check",
+			Usage: "check that the sudoers file is up-to-date with ~/.lima/_config/networks.yaml",
+		},
+	},
+	Action: sudoersAction,
+}
+
+func sudoersAction(clicontext *cli.Context) error {
```
Comment (AkihiroSuda) [inline cmd/limactl/sudoers.go:23]: error out when GOOS != "darwin"

File: pkg/networks/commands.go (hunk 3 above)
Comment (jandubois) [reply pkg/networks/commands.go:68]: Yes, I intended to do this as part of `some validation for networks.yaml`. Not sure about the dylibs though; I don't want to shell out to `otool` to look at the import paths, and then figure out how to resolve `@executable_path/...` references etc. Let me know how you would want to determine the library dependencies!

I was planning to make sure all `config.Path` entries don't have any whitespace, and the files and all parent etc. directories are only writable by root.

Comment (AkihiroSuda) [reply pkg/networks/commands.go:68]: The otool stuff can be another PR in future.

File: pkg/limayaml/validate.go (hunk 1 above)
Comment (jandubois) [reply pkg/limayaml/validate.go:181]: Ok, how do you want to implement it. As a separate list under `network`, e.g.

```yaml
network:
  lima:
  - shared
  - host
```

File: pkg/networks/commands.go (hunk 2 above)
Comment (jandubois) [reply pkg/networks/commands.go:9]: I would add support for PTP mode in a later PR; I believe it can be added as just another network, albeit one that can only be used by one instance at a time.

And e.g. a host-only network doesn't work in PTP mode because you can't connect multiple instances to it. Maybe it doesn't make sense to consider this, but I was thinking about being able have one instance on a host-only network, and a second instance running as a proxy server/jumpbox, to test that an application works correctly in a fully proxied environment.

File: pkg/limayaml/validate.go (hunk 1 above)
Comment (AkihiroSuda) [reply pkg/limayaml/validate.go:181]: SGTM

Comment (jandubois) [reply pkg/limayaml/validate.go:181]: It will have to be:

```yaml
network:
  lima:
  - network: bridged
    macAddress: 11:22:33:44:55:66
    name: lima0
  - network: host
```

etc. It is unfortunate that we use `vde.name` for the interface name. I wish it was `interface`, so we could use `lima.name` and `lima.interface` instead of `lima.network` and `lima.name`, but I guess it is too late to change this now.

Any strong feelings about the field names?

Comment (AkihiroSuda) [reply pkg/limayaml/validate.go:181]: I don't have a strong opinion.

If we want to change `name` to `interface`, we can safely do that by deprecating `name` with a warning message.

Comment (jandubois) [reply pkg/limayaml/validate.go:181]: This turns out to be really painful to implement, as we would have to duplicate so much code in validation. And the code would still not be independent as we would still have to verify that there are no duplicate interface names between the `VDE` and the `lima` lists.

So if we can still change the `network` layout, then I would suggest going with:

```yaml
network:
- vnl: /var/run/vde.ctl
  interface: vde0
- lima: shared
  interface: lima0
```

Either `vnl` or `lima` would be required, and `switchPort` would only be valid when you use `vnl`.

The problem is that we can't really do a proper deprecation message, as `network` would turn from a `map` to a `slice`.

I still think reserving the `lima://` scheme for our internal use is the most elegant solution, as I think it will be rather unlikely that there ever will be a VDE4 plugin called `lima`. But if we can change the `network` layout, then I'm fine with that too.

Comment (jandubois) [reply pkg/limayaml/validate.go:181]: Ok, one more thought: we could call the new structure `networks`, and deprecate `network`.

Comment (jandubois) [reply pkg/limayaml/validate.go]: > we could call the new structure `networks`, and deprecate `network`.

I've implemented this in https://github.com/lima-vm/lima/pull/239/commits/03e90a2 (separate commit for review; will squash later).

It throws an error when you use both the new and the deprecated fields:

```
$ l start alpine.yaml
? Creating an instance "alpine" Proceed with the default configuration
FATA[0001] the YAML is invalid, saved the buffer as "lima.REJECTED.yaml": you cannot use deprecated field `network.VDE` together with replacement field `networks`
```

And will show a deprecation warning when the file uses the old field:

```
$ l start alpine.yaml
? Creating an instance "alpine" Proceed with the default configuration
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] field `network.VDE` is deprecated; please use `networks` instead
WARN[0001] public key "/Users/jan/.ssh/foo.pub" doesn't seem to be in ssh format
INFO[0001] Attempting to download the image from "https://github.com/lima-vm/alpine-lima/releases/download/v0.1.4/alpine-lima-std-3.13.5-x86_64.iso"
...
```

The warning is show multiple times because the file is validated several times. I'm somewhat surprised it happens 5 times, but I think it is unavoidable to do it multiple times, unless we cache all the instance data returned by `store.Inspect`.

Haven't done the other changes (network validation, metrics, help), but let me know if you think this is now an acceptable way to declare the networks for an instance!

This is now a working minimal network config:

```yaml
networks:
- lima: shared
```

And I changed the default interface names from `vde0` etc. to `lima0` etc because the use of vde is kind of an implementation details now.

File: cmd/limactl/sudoers.go (hunk 4 above)
Comment (jandubois) [reply cmd/limactl/sudoers.go:25]: > error out when GOOS != "darwin"

Done in 97d3a2b9256ab5503ab8176a1741deea48cfb6f8

File: pkg/networks/commands.go (hunk 3 above)
Comment (jandubois) [reply pkg/networks/commands.go:68]: > Can we error out when the `vde_vmnet` binary (or its DLLs) is modifiable by non-root?

Done in b6bcc80ba862b348839f62842bee9dd019550dec

(otool logic to come in a later PR)

File: cmd/limactl/sudoers.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		var file string
+		switch clicontext.NArg() {
+		case 0:
+			config, err := networks.Config()
+			if err != nil {
+				return err
+			}
+			file = config.Paths.Sudoers
+			if file == "" {
+				return errors.New("no sudoers file defined in ~/.lima/_config/networks.yaml")
```
Comment (AkihiroSuda) [inline cmd/limactl/sudoers.go:39]: This message would be confusing when LIMA_HOME is set to a non-default path

File: pkg/networks/commands.go (hunk 3 above)
Comment (jandubois) [reply pkg/networks/commands.go:68]: > (otool logic to come in a later PR)

I found that `/usr/bin/otool` is just a shim that tries to locate `XCode.app`, and if it can't find it presents a dialog offering to install the "commandline developer tools".

So it is going to be tricky to figure out if it is even safe to call `otool`. And what do we do when it is not available? Fall back on checking the "known filenames"?

Comment (AkihiroSuda) [reply pkg/networks/commands.go:68]: Maybe we can parse macho directly (in another PR)

Comment (AkihiroSuda) [reply pkg/networks/commands.go:68]: This should work https://pkg.go.dev/debug/macho#File.ImportedLibraries

**Comment by AkihiroSuda** [comment]
What are the remaining tasks to move this out of draft?

**Comment by jandubois** [comment]
> What are the remaining tasks to move this out of draft?

Things that I think *must* be done as part of this PR are:

* improve `networks.Validate`:
  *  currently allows all paths to be writable by `daemon`; only `varRun` should be writable
  * the lowest level parent directory of `varRun` that already exists *must* be writable by `daemon`, so the `varRun` directory can be created
* update docs
  * document `$LIMA_HOME/_config/networks.yaml`
  * document `$LIMA_HOME/_networks`

One thing I would like to still address in the PR:

* `lima.yaml` is validated 3-5 times, which shows any warnings just as often. Either reduce the number of calls, or cache the results.

Things that can be done in another PR:

* check dylib dependencies of `vde_*` binaries to be only root-writable
* set interface metrics so that the vde interfaces will be prioritized over slirp for outgoing connections
* add support for PTP networks

Things that must be a separate issue/PR:

* managed network support for Linux (using `tap`).

**Comment by jandubois** [comment]
> * `lima.yaml` is validated 3-5 times, which shows any warnings just as often. Either reduce the number of calls, or cache the results.

Haven't found a non-convoluted way of doing this, so I added a `warn` parameter to `limayaml.Validate` to only issue warnings when called by directly from the `start` command. Validation is quick, so it doesn't hurt to call it multiple times.

I'm going to move the PR out of draft now; I think everything else can wait for follow-up PRs.

@AkihiroSuda Do you want me to squash everything to a single commit, or do you want to keep anything separate (e.g. the re-arranging of packages to prevent cyclical import dependencies). I do not want to re-arrange the order of commits though, I would rather squash everything into a single mega-commit.

File: pkg/networks/config.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		if cache.err != nil {
+			return
+		}
+		_, cache.err = os.Stat(configFile)
+		if cache.err != nil {
+			if !errors.Is(cache.err, os.ErrNotExist) {
+				return
+			}
+			configDir := filepath.Dir(configFile)
+			cache.err = os.MkdirAll(configDir, 0700)
```
Comment (jandubois) [inline pkg/networks/config.go:48]: Not sure why I picked `0700`; it probably should be `0755`.

File: pkg/qemu/qemu.go [hunk 7]
```diff
@@ -257,33 +258,49 @@ func Cmdline(cfg Config) (string, []string, error) {
 
 	// Network
 	args = append(args, "-netdev", fmt.Sprintf("user,id=net0,net=%s,dhcpstart=%s,hostfwd=tcp:127.0.0.1:%d-:22",
-		qemuconst.SlirpNetwork, qemuconst.SlirpIPAddress, y.SSH.LocalPort))
+		qemu.SlirpNetwork, qemu.SlirpIPAddress, y.SSH.LocalPort))
 	args = append(args, "-device", "virtio-net-pci,netdev=net0,mac="+limayaml.MACAddress(cfg.InstanceDir))
-	if len(y.Network.VDE) > 0 && !strings.Contains(string(features.NetdevHelp), "vde") {
+	usingVDE := false
+	for _, nw := range y.Networks {
+		if nw.VNL != "" {
```
Comment (jandubois) [inline pkg/qemu/qemu.go:265]: Oops, that was a thinko, the managed networks also require `netdev "vde"` support.

File: pkg/networks/sudoers.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+		sb.WriteString(fmt.Sprintf("# Manage %q network daemons\n", name))
+		for _, daemon := range []string{Switch, VMNet} {
+			sb.WriteRune('\n')
+			sb.WriteString(fmt.Sprintf("%%%s ALL=(%s:%s) NOPASSWD:NOSETENV: \\\n",
+				config.Group, config.DaemonUser(daemon), config.DaemonGroup(daemon)))
+			sb.WriteString(fmt.Sprintf("    %s, \\\n", config.StartCmd(name, daemon)))
+			sb.WriteString(fmt.Sprintf("    %s\n", config.StopCmd(name, daemon)))
+		}
+	}
+	return sb.String(), nil
```
Comment (AkihiroSuda) [inline pkg/networks/sudoers.go:41]: We should have unit tests to validate the generated strings.
Can be another PR, though.

File: pkg/networks/validate.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+			return findBaseDirectory(filepath.Dir(path))
+		}
+	}
+	return path
+}
+
+const (
+	RootUID   = 0
+	WheelGID  = 0
+	DaemonGID = 1
```
Comment (AkihiroSuda) [inline pkg/networks/validate.go:54]: These should be moved to `*_darwin.go`

File: pkg/limayaml/validate.go [hunk 10]
```diff
@@ -12,12 +12,13 @@ import (
...
 	"github.com/lima-vm/lima/pkg/localpathutil"
+	"github.com/lima-vm/lima/pkg/networks"
 	"github.com/lima-vm/lima/pkg/osutil"
-	"github.com/lima-vm/lima/pkg/qemu/qemuconst"
+	"github.com/lima-vm/lima/pkg/qemu/const"
 	"github.com/sirupsen/logrus"
 )
 
-func Validate(y LimaYAML) error {
+func Validate(y LimaYAML, warn bool) error {
```
Comment (AkihiroSuda) [inline pkg/limayaml/validate.go:21]: Maybe we should make this like `Validate(y LimaYAML)(warns []error, err error)

**Comment by AkihiroSuda** [comment]
> Do you want me to squash everything to a single commit, or do you want to keep anything separate (e.g. the re-arranging of packages to prevent cyclical import dependencies). I do not want to re-arrange the order of commits though, I would rather squash everything into a single mega-commit.

Either is fine to me

File: pkg/networks/validate.go (hunk 9 above)
Comment (jandubois) [reply pkg/networks/validate.go:54]: > These should be moved to `*_darwin.go`

Just those constants, or the whole code?

The logic is not really darwin specific; it is just that we only call it on darwin.

Even the constants are kind of universal; except that "wheel" should be called "root" on Linux. I'm fine with using `RootGID`. So if anything we should have platform specific constants for the string values (e.g. `RootGroup = "wheel"` on darwin).

Comment (AkihiroSuda) [reply pkg/networks/validate.go:54]: > Just those constants, or the whole code?

Either is fine.

At least `DaemonGID=1` is specific to Darwin.
On Linux, the daemon GID is often 2.

**Comment by AkihiroSuda** [comment]
Could you update `examples/vmnet.yaml` to the new syntax to test this on CI?

File: pkg/limayaml/validate.go (hunk 10 above)
Comment (jandubois) [reply pkg/limayaml/validate.go:21]: > Maybe we should make this like `Validate(y LimaYAML)(warns []error, err error)`

We could, but I'm not keen on it: there are over 50 `return` statements that would all need to deal with it:

```
$ grep return pkg/limayaml/validate.go | wc -l
      54
```

Just for the benefit of 3 possible warnings:

```
$ grep 'logrus\.' pkg/limayaml/validate.go
				logrus.Warnf("field `network.VDE` is deprecated; please use `networks` instead")
					logrus.WithError(err).Debugf("field `%s.vnl` %q failed stat", field, vdeSwitch)
					logrus.Warnf("field `%s.vnl` is unlikely to work for %s (unless libvdeplug4 has been ported to %s and is installed)",
```

And the warnings don't even come from `Validate` directly, but from `validateNetworks`.

I'll do it if you really think it makes a difference, but I think the code is clearer the way it currently is.

File: pkg/networks/validate.go (hunk 9 above)
Comment (jandubois) [reply pkg/networks/validate.go]: I haven't split the code into platform-specific pieces, but made it look all user and group names / id dynamically: https://github.com/lima-vm/lima/pull/239/commits/c96f16b802ae8cf3566761717ca423cbbc01d6ef

File: pkg/networks/sudoers.go (hunk 8 above)
Comment (jandubois) [reply pkg/networks/sudoers.go:45]: > We should have unit tests to validate the generated strings.

Added in https://github.com/lima-vm/lima/pull/239/commits/42fcd35d4ad0bee35253f089d728d20f7a9f42d4

**Review by AkihiroSuda — APPROVED** [review]
Thanks
